### PR TITLE
Complete #91

### PR DESCRIPTION
**Pull request by NovaMachina**
This creates the domain for the workspace cache system. The WorkspaceTree is abstract to accept any type for the Nodes. Package, SourceFile, and BuildTarget are specific domain objects for Bazel.

### CODE REVIEW

**Review by loganmcbride148 — APPROVED** [review]
Looks good to me, I'm not sure if you've already tested this or not, but I imagine it's not that hard to test it in integration.

File: server/src/main/java/server/cache/BUILD
Comment (josiahsrc) [inline server/src/main/java/server/cache/BUILD]: See comments below, but perhaps we ought to move all this code into the `server.bazel` package?

File: server/src/main/java/server/cache/WorkspaceTree.java
Comment (josiahsrc) [inline server/src/main/java/server/cache/WorkspaceTree.java]: What's the purpose of the generics here?

File: server/src/main/java/server/cache/BuildTarget.java
Comment (josiahsrc) [inline server/src/main/java/server/cache/BuildTarget.java]: This looks like an object that the Bazel package should also know about. E.g. when we call `BazelInvoker.getBuildTargets(...)`, we might want to return this type instead of duplicating types. Maybe we ought to move this `server.cache` package to be the actual `server.bazel` or possibly `server.bazel.domain` package?

File: server/src/main/java/server/cache/Package.java
Comment (josiahsrc) [inline server/src/main/java/server/cache/Package.java]: Same comment as above, this is bazel-specific domain

**Review by josiahsrc — COMMENTED** [review]
Looks really great. I have a few things which I think are worth discussing which I added in my review. Lmk your thoughts

File: server/src/main/java/server/cache/WorkspaceTree.java
Comment (NovaMachina) [reply server/src/main/java/server/cache/WorkspaceTree.java]: Just good practices.

**Comment by NovaMachina** [comment]
As per @josiahsrc suggestions. I've changed the locations of where these files are located.

File: server/src/main/java/server/cache/WorkspaceTree.java
Comment (josiahsrc) [reply server/src/main/java/server/cache/WorkspaceTree.java]: I mean in terms of usage. The generic in this case would restrict all nodes in the tree to be of type `T`. How will that tie in to querying and constructing the tree? Our tree will need to look something like this

```
Workspace(
  children: [
    BuildFile(
      children: [
        Target(),
        Target(),
      ],
    ),
     BuildFile(
      children: [
        Target(),
        Target(),
      ],
    ),
  ],
)
```

But the Workspace is constrained to nodes of type T only, so it doesn't seem like we can provide this structure?

Comment (NovaMachina) [reply server/src/main/java/server/cache/WorkspaceTree.java]: If you plug `Package` in for `T` then it works.

File: server/src/main/java/server/bazel/WorkspaceTree.java
Comment (josiahsrc) [inline server/src/main/java/server/bazel/WorkspaceTree.java]: This seems like it's also domain, but someone fight me on this

File: server/src/main/java/server/cache/WorkspaceTree.java
Comment (josiahsrc) [reply server/src/main/java/server/cache/WorkspaceTree.java]: So why not use package instead of T?

File: server/src/main/java/server/bazel/WorkspaceTree.java
Comment (NovaMachina) [reply server/src/main/java/server/bazel/WorkspaceTree.java]: I mean, its more of a data structure for the cache rather than a domain object like `Package` or `SourceFile`. That was my thinking.

Comment (josiahsrc) [reply server/src/main/java/server/bazel/WorkspaceTree.java]: It seems like it's all part of the same data structure to me, because we'll have querying that looks like this ish

```
Workspace.findPackage(...).findTarget(...)
```

In this case, both the `WorkspaceTree` and `Package` are queryable. Maybe it ought to all go in a package called `server.bazel.tree`? That way its all in one spot, reusable, and clear that it's bazel domain

File: server/src/main/java/server/cache/WorkspaceTree.java
Comment (NovaMachina) [reply server/src/main/java/server/cache/WorkspaceTree.java]: What if we wanted a Tree of SourceFiles for some reason?

File: server/src/main/java/server/bazel/WorkspaceTree.java
Comment (NovaMachina) [reply server/src/main/java/server/bazel/WorkspaceTree.java]: I like that. Moved it there.

File: server/src/main/java/server/cache/WorkspaceTree.java
Comment (josiahsrc) [reply server/src/main/java/server/cache/WorkspaceTree.java]: Point ☝️  In that case we will have to generalize the queries to work for either packages or source files. @jasonxris This is going to impact your task, what do you think here?

Another thing to consider is that bazel workspaces are techincally composed of packages, source files, and targets, all in the same tree. Representing this all by one WorkspaceTree object might be easier to understand in the context of bazel

Comment (NovaMachina) [reply server/src/main/java/server/cache/WorkspaceTree.java]: `Package` contain a list of `BuildTarget` and `SourceFile`. So creating a `WorkspaceTree<Package>` will create a tree object that has a Bazel context.

Comment (josiahsrc) [reply server/src/main/java/server/cache/WorkspaceTree.java]: My main concern is that generalizing a <T> param for a workspace isn't needed. This Workspace tree is only meant for Bazel, we don't have any other uses for it. So why generalize?

If we generalize now, the api for querying a workspace will become more complicated to implement, extend, maintain, and invoke. Reason being is because we will be forced to generalize *how* we query. For example, if we want the WorkspaceTree to be queryable for packages, packages would likely have to implement some interface `IQueryable` so that when calling the workspace we could do something like

```
WorkspaceTree.findByString(<Some generalized type to query by>)

...

class Package implements IQueryable<String>, IQueryable<int>, IQueryable<BazelRule>, ... {
  ...
}
```

Point being that generalizing isn't necessary here. It will only make our lives harder. In my experience, you should **only** generalize if you have a legitimate need for it.

**Review by josiahsrc — APPROVED** [review]
Let's get @jasonxris's input before merging this in. He'll be working with this implementation

**Comment by jasonxris** [comment]
@NovaMachina 
Looks really solid. 

1. Might be worth adding a toString or a outputExplicitPath method to the build target object that would output the explicit target path. (path + ":" + label). 

2. It might be worth adding another variable to the source model to separate the name from the file Extension (myFile.java would be stored as filename = "myFile" and sourceExtension=".java").  I could see this being useful in the future when we want the source files to match the rule type. For example a java_binary rule should only autocomplete .java files when inputting sources from the package.

**Comment by NovaMachina** [comment]
@jasonxris I added what you suggested. Can you review and then approve if that is what you were thinking?

**Review by jasonxris — APPROVED** [review]
Looks good.

**Pull request merged**
